chain/trie/pruning.go

```diff
@@ -0,0 +1,305 @@
+package trie
+
+import (
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/nknorg/nkn/chain/db"
+	"github.com/nknorg/nkn/common"
+)
+
+type RefCounts struct {
+	trie   *Trie
+	counts map[common.Uint256]uint32
+
+	startRefCountHeight  uint32
+	startPruningHeight   uint32
+	targetRefCountHeight uint32
+	targetPruningHeight  uint32
+
+	dbCounter   int32
+	nodeCounter int32
+}
+
+func NewRefCounts(t *Trie, refCountTargetHeight, pruningTargetHeight uint32) *RefCounts {
+	ref := &RefCounts{
+		counts:               make(map[common.Uint256]uint32, 0),
+		trie:                 t,
+		targetRefCountHeight: refCountTargetHeight,
+		targetPruningHeight:  pruningTargetHeight,
+	}
+
+	heightBuffer, err := t.db.Get(db.TriePrunedHeightKey())
+	if err != nil {
+		ref.startPruningHeight = 0
```

Comment:
Suggest to add an info log here and below to help debugging in case there is a real error